optimum/onnxruntime/optimization.py

```diff
@@ -143,20 +143,30 @@ def optimize(
         LOGGER.info("Optimizing model...")
 
         for model_path in self.onnx_model_path:
-            optimizer = optimize_model(
-                model_path.as_posix(),
-                model_type,
-                self.normalized_config.num_attention_heads,
-                self.normalized_config.hidden_size,
-                opt_level=optimization_config.optimization_level,
-                optimization_options=optimization_options,
-                use_gpu=optimization_config.optimize_for_gpu,
-                only_onnxruntime=not optimization_config.enable_transformers_specific_optimizations,
-            )
-
-            if optimization_config.fp16:
-                # keep_io_types to keep inputs/outputs as float32
-                optimizer.convert_float_to_float16(keep_io_types=True)
+            try:
+                optimizer = optimize_model(
+                    model_path.as_posix(),
+                    model_type,
+                    self.normalized_config.num_attention_heads,
+                    self.normalized_config.hidden_size,
+                    opt_level=optimization_config.optimization_level,
+                    optimization_options=optimization_options,
+                    use_gpu=optimization_config.optimize_for_gpu,
+                    only_onnxruntime=not optimization_config.enable_transformers_specific_optimizations,
+                )
+
+                if optimization_config.fp16:
+                    # keep_io_types to keep inputs/outputs as float32
+                    optimizer.convert_float_to_float16(
+                        use_symbolic_shape_infer=not optimization_config.disable_shape_inference, keep_io_types=True
+                    )
+            except Exception as e:
+                if "Incomplete symbolic shape inference" in str(e):
+                    err = RuntimeError(
+                        f"{str(e)}. Try to set `disable_shape_inference=True` in your optimization configuration."
+                    )
+                    raise err from e
```

Comment:
How does it compare to:
```
raise RuntimeError(f"{str(e)}. Try to set `disable_shape_inference=True` in your optimization configuration.")
```

?

Reply:
The current solution returns:
```
Traceback (most recent call last):
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 160, in optimize
    optimizer.convert_float_to_float16(
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../onnx_model.py", line 594, in convert_float_to_float16
    model = shape_infer_helper.infer_shapes(model, auto_merge=True, guess_output_rank=False)
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../../tools/symbolic_shape_infer.py", line 2347, in infer_shapes
    raise Exception("Incomplete symbolic shape inference")
Exception: Incomplete symbolic shape inference

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_optimum.py", line 26, in <module>
    optimizer.optimize(save_dir=onnx_path, optimization_config=optimization_config)
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 168, in optimize
    raise err from e
RuntimeError: Incomplete symbolic shape inference. Try to set `disable_shape_inference=True` in your optimization configuration.
```

While raising directly the error returns:
```
Traceback (most recent call last):
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 160, in optimize
    optimizer.convert_float_to_float16(
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../onnx_model.py", line 594, in convert_float_to_float16
    model = shape_infer_helper.infer_shapes(model, auto_merge=True, guess_output_rank=False)
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../../tools/symbolic_shape_infer.py", line 2347, in infer_shapes
    raise Exception("Incomplete symbolic shape inference")
Exception: Incomplete symbolic shape inference

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_optimum.py", line 26, in <module>
    optimizer.optimize(save_dir=onnx_path, optimization_config=optimization_config)
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 170, in optimize
    raise e
RuntimeError: Incomplete symbolic shape inference. Try to set `disable_shape_inference=True` in your optimization configuration.
```

I think the message `The above exception was the direct cause of the following exception:` makes things less confusing.

I also thought about directly modifying the message of the exception raised by ORT but I think it's better to have a clear separation between ORT and Optimum error messages.

Reply:
Oh I see, yes agreed!